design-proposals/klip-49-add-source-stream-table-semantic.md

```diff
@@ -0,0 +1,161 @@
+## High Level Design
+
+### Semantic Change
+We shall first add optional keyword `SOURCE` into the KSQL codebase, by adding the new syntax to `sqlBase.g4` for:
+* createStream
+* createTable
+
+with the sample syntax as:
+```
+CREATE (OR REPLACE)? (SOURCE)? TABLE (IF NOT EXISTS)?...
+```
+
+### Make Source Stream/Table Read-only
+When doing insertions, KSQL would check whether the given stream/table is read-only. If so, the insertion will be rejected.
+
+Additionally, when calling `DROP` command on a source stream/table, currently user could do `DROP STREAM/TABLE 
+[table_name] DELETE TOPIC` to clean up the underlying topic backed by the stream/table. This should not be allowed for 
+source stream/table as the entity has no ownership of the input data. See this 
+[issue](https://github.com/confluentinc/ksql/issues/3585) for more context. 
+
+### Source Table Materialization
+The main difference between a source table and a materialized state by a CTAS query is that we do not populate a sink topic 
+for source table, as it is not expected to have a downstream connection. This will be addressed when constructing the 
+`CREATE SOURCE TABLE` as a special type of persistent query, which would be an implementation detail. 
+
+On the other hand, for the first version of release, we could propagate that `source table` serves as a syntax sugar 
+for `create table; create table as select *` in a minimum, so that in the long run users would feel less surprised when 
+they start seeing table materialization as default. In the future as we have higher persistent query limit and 
+introduce materialized views and complete the ownership model, we can nudge our messaging to focus on the ownership 
+side, and perhaps also allow users to opt-out of materialization if they want to reduce footprint.
+
+Another consideration we have is how to expose such type of query to the end users. Generally speaking, any active 
+stream runtime should be represented as a query to present to end users. For source table materialization, it is a 
+stateful job that comes with overhead, which should definitely be exposed to end users. However, any exposed query 
+could be potentially terminated directly by the end user, which comes with a consistency problem with source table 
+semantic agreement. If a source table query could be terminated independently by the user, the original source table 
+would be dangling and not available for pull query anymore. To address this problem, we have two approaches:
+ 
+ 1. The KSQL server rejects the termination of a source table query and informs end user to try deleting the original source table.
+ 2. The KSQL server will verify the deletion of a query, and try deleting the associated source table with it.
+ 
+Both approaches are valid here to maintain the consistency, but considering when dropping a source table, the active 
+materialization query would be terminated, I'm inclined to suggest taking approach #1 here for simplicity, and avoid 
```

Comment:
So just to clarify: the plan is to show the "query" associated with source table materialization as part of `LIST QUERIES;`, but to reject an attempt to terminate the query? Previously we'd discussed not showing a query ID for these queries at all, so users cannot even try to terminate them. I have a slight preference for the latter but am fine either way.